DataStructures/AATree/AATree.cs

```diff
@@ -0,0 +1,587 @@
+        /// Useful for debugging.
+        /// </summary>
+        /// <remarks>
+        /// The properties that are checked are:
+        /// <list type="number">
+        /// <item>The level of every leaf node is one.</item>
+        /// <item>The level of every left child is exactly one less than that of its parent.</item>
+        /// <item>The level of every right child is equal to or one less than that of its parent.</item>
+        /// <item>The level of every right grandchild is strictly less than that of its grandparent.</item>
+        /// <item>Every node of level greater than one has two children.</item>
+        /// </list>
+        /// More information: https://en.wikipedia.org/wiki/AA_tree .
+        /// </remarks>
+        /// <param name="node">The node to check from.</param>
+        /// <returns>true if node passes all checks, false otherwise.</returns>
+        private bool ValidateTree(AATreeNode<TKey>? node)
+        {
+            if (node == null)
+            {
+                return true;
+            }
+
+            // Check level == 1 if node if a leaf node.
+            bool leafNodeCheck = CheckLeafNode(node);
+
+            // Check level of left child is exactly one less than parent.
+            bool leftCheck = CheckLeftSubtree(node);
+
+            // Check level of right child is equal or one less than parent.
+            bool rightCheck = CheckRightSubtree(node);
+
+            // Check right grandchild level is less than node.
+            bool grandchildCheck = CheckRightGrandChild(node);
+
+            // Check if node has two children if not leaf.
+            bool nonLeafChildrenCheck = CheckNonLeafChildren(node);
+
+            bool thisNodeResult = leafNodeCheck &&
+                leftCheck &&
+                rightCheck &&
+                grandchildCheck &&
+                nonLeafChildrenCheck;
+
+            return thisNodeResult && ValidateTree(node.Left) && ValidateTree(node.Right);
+        }
+
+        /// <summary>
+        /// Checks if node is a leaf, and if so if its level is 1.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckLeafNode(AATreeNode<TKey> node)
+        {
+            if (node.Left == null && node.Right == null &&
+                node.Level != 1)
+            {
+                return false;
+            }
+            else
+            {
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Checks if left node's level is exactly one less than node's level.
+        /// </summary>
+        /// <param name="node">The node to check.</param>
+        /// <returns>true if node passes check, false otherwise.</returns>
+        private bool CheckLeftSubtree(AATreeNode<TKey> node)
+        {
+            if (node.Left != null && node.Level - node.Left.Level != 1)
```

Comment:
Codacy found an issue: [Use the '?:' operator here.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6367315)